lib/resty/wasmx/shm.lua

```diff
@@ -0,0 +1,577 @@
+    end
+
+    local rc = C.ngx_wa_ffi_shm_metric_record(metric_id, value)
+    if rc == FFI_DECLINED then
+        return nil, "metric not found"
+    end
+
+    if rc ~= FFI_OK then
+        return nil, "unknown error"
+    end
+
+    return true
+end
+
+
+local function parse_cmetric(cmetric)
+    if cmetric.metric_type == _types.ffi_metric.COUNTER then
+        return {
+            type = "counter",
+            value = tonumber(cmetric.slots[0].counter),
+        }
+    end
+
+    if cmetric.metric_type == _types.ffi_metric.GAUGE then
+        return {
+            type = "gauge",
+            value = tonumber(cmetric.slots[0].gauge.value),
+        }
+    end
+
+    if cmetric.metric_type == _types.ffi_metric.HISTOGRAM then
+        local hbuf = cmetric.slots[0].histogram
+        local ch = ffi_cast("ngx_wa_metrics_histogram_t *", hbuf)
+        local h = { type = "histogram", value = {} }
+
+        for i = 0, ch.n_bins do
+            local cb = ch.bins[i]
+            if cb.upper_bound == 0 then
+                break
+            end
+
+            insert(h.value, {
+                ub = cb.upper_bound,
+                count = cb.count,
+            })
+        end
+
+        return h
+    end
+
+    assert(false, "unreachable")
+end
+
+
+local function metrics_get_by_id(zone, metric_id)
+    if type(metric_id) ~= "number" then
+        error("metric_id must be a number", 2)
+    end
+
+    ffi_fill(_mbuf, _mbs)
+    ffi_fill(_hbuf, _hbs)
+
+    local rc = C.ngx_wa_ffi_shm_metric_get(metric_id, nil,
+                                           _mbuf, _mbs,
+                                           _hbuf, _hbs)
+    if rc == FFI_DECLINED then
+        return nil
+    end
+
+    if rc ~= FFI_OK then
+        return nil, "unknown error"
+    end
+
+    return parse_cmetric(ffi_cast("ngx_wa_metric_t *", _mbuf))
+end
+
+
+---
+-- ngx_wasm_module internally prefixes metric names according to
+-- where they have been defined, e.g. "pw.filter.*", "lua.*", or
+-- "wa.*".
+--
+-- metrics_get_by_name assumes it is retrieving a Lua-defined metric
+-- and will by default prefix the given name with `lua.`
+--
+-- This behavior can be disabled by passing `opts.prefix` as false.
+local function metrics_get_by_name(zone, name, opts)
+    if type(name) ~= "string" or name == "" then
+        error("name must be a non-empty string", 2)
+    end
+
+    if opts ~= nil then
+        if type(opts) ~= "table" then
+            error("opts must be a table", 2)
+        end
+
+        if opts.prefix ~= nil and type(opts.prefix) ~= "boolean" then
+            error("opts.prefix must be a boolean", 2)
+        end
+    end
+
+    ffi_fill(_mbuf, _mbs)
+    ffi_fill(_hbuf, _hbs)
+
+    name = (opts and opts.prefix == false) and name or "lua." .. name
+
+    local cname = ffi_new("ngx_str_t", { data = name, len = #name })
+
+    local rc = C.ngx_wa_ffi_shm_metric_get(0, cname, _mbuf, _mbs, _hbuf, _hbs)
+    if rc == FFI_DECLINED then
+        return nil
```

Comment:
```suggestion
        return nil, `metric not found`
```

Reply:
This is a single return value API just like shm `get()`.